File: scripts/py/bible_reader_words_known.py
"""Which of these words the reader already had an answer for before any Bible name was filed.

The reader ships with a dictionary of its own, and the Bible names are filed on
top of it.  Two of the three sets of Bible names are filed only where the reader
had nothing - a derived reading of an ordinary word must not be forced on a
reader that already has one - and the third, the names somebody wrote down, is
filed over the top of whatever was there.

★ THIS ASKS WHETHER THAT THIRD SET IS OVERWRITING ANSWERS OR FILLING GAPS, WHICH
IS A DIFFERENT QUESTION FROM WHETHER ITS ANSWERS ARE GOOD.  Filling a gap can
only improve on a name sounded out letter by letter.  Overwriting is a claim that
the Bible lexicon knows a word better than the reader's own dictionary does, and
that claim is worth exactly as much as the entry behind it - measured, "Israel"
was already read the way the standard dictionary reads it and was overwritten
with a reading no dictionary records.

The reader is asked before anything is filed on top of it, which is the only
moment the question can be asked at all: afterwards every name is known, because
filing it is what made it known.
"""

import json
import sys

from text_to_speech import g2p_plain


def main(args_path):
    with open(args_path, encoding="utf-8") as fh:
        args = json.load(fh)
    words = args["words"]
    g2p = g2p_plain()
    golds = set(g2p.lexicon.golds)
    silvers = set(getattr(g2p.lexicon, "silvers", {}))
    known = golds | silvers
    answered = [w for w in words if w in known]
    unanswered = [w for w in words if w not in known]
    said = {}
    for word in answered:
        if word in g2p.lexicon.golds:
            said[word] = g2p.lexicon.golds[word]
        else:
            said[word] = g2p.lexicon.silvers[word]
    print(
        json.dumps(
            {
                "words_asked": len(words),
                "reader_golds": len(golds),
                "reader_silvers": len(silvers),
                "answered_already": answered,
                "no_answer": unanswered,
                "reader_says": said,
            },
            ensure_ascii=False,
        )
    )


if __name__ == "__main__":
    main(sys.argv[1])

File: js/bible_reader_words_known.mjs
import { arguments_assert } from "./arguments_assert.mjs";
import { text_split_comma_or_empty } from "./text_split_comma_or_empty.mjs";
import { fn_name } from "./fn_name.mjs";
import { py_script_speech_json_report } from "./py_script_speech_json_report.mjs";
export async function bible_reader_words_known(words) {
  "$plain words";
  "Splits the words named into the ones the reader already had an answer for and the ones it had none for, asked of the reader before any Bible name is filed on top of it.";
  "★ IT SEPARATES A GAP BEING FILLED FROM AN ANSWER BEING OVERWRITTEN, WHICH ARE WORTH OPPOSITE AMOUNTS. A Bible name filed where the reader had nothing can only improve on letters sounded out one at a time. The same name filed over an answer the reader already had is a claim to know better than the standard dictionary, and that claim has to be right word by word rather than in general.";
  "★ THE READER IS ASKED BEFORE THE FILING, WHICH IS THE ONLY MOMENT THE QUESTION EXISTS. Afterwards every Bible name is known, because filing it is what made it known, and the answer would be yes to everything.";
  "The words arrive as one comma-joined word, as every list does on a command line, and are split here rather than at the far end, where a word handed over whole is read letter by letter and quietly matches nothing.";
  arguments_assert(arguments, 1);
  let asked = text_split_comma_or_empty(words);
  let script_name = fn_name("bible_reader_words_known");
  let args = {
    words: asked,
  };
  let reported = await py_script_speech_json_report(script_name, args);
  return reported;
}
